### DISCUSSION TITLE
[Libmesh-users] Problem with parallel execution of element loop

### DISCUSSION THREAD

**Discussion by jwpeterson** [libMesh/libmesh#3086] [Q&A] [answered]
> The following is from an [email](https://sourceforge.net/p/libmesh/mailman/message/37382730/) posted to the libmesh-users mailing list.

The following code fragment is part of a program that behaves correctly when invoked to run on a single processor (ie using : mpirun -n 1 main)
```
    libMesh::BoundaryInfo& bdry_info = tet_mesh.get_boundary_info();
    for (auto pp_elem = tet_mesh.local_elements_begin(); pp_elem != tet_mesh.local_elements_end(); ++pp_elem)
    {
        Elem* elem = *pp_elem;
        for( unsigned int is = 0; is < 4; ++is)
            if ( /* stuff */)
                bdry_info.add_side(elem, is, sideset_id);
    }
```
However, when I invoke it with several processors (ie using mpirun -n 4 main), the BoundaryInfo object does not contain all of the sidesets that it contains on a single processor.

I suspect that the issue here is that several processors are writing to the same memory location (ie the BoundaryInfo object) without appropriate synchonization/mutex code.

I certainly didn't add any such code.

Is my suspicion correct or are there other problems with the code?

In the former case, can someone point me to some correct code in which mesh elements are processed in parallel and a common data structure is modified by each thread/process?

dan

**Comment by jwpeterson** [comment, marked as answer]
I think you are seeing the expected behavior since you are looping over "local" elements. The `BoundaryInfo` object is not shared by the processors... there is a separate copy of the BoundaryInfo object on each processor and each will contain only boundary info about local elements after running the loop above. The BoundaryInfo class also provides the
```
  void parallel_sync_side_ids();
  void parallel_sync_node_ids();
```
APIs for syncing this data across processors.

**Reply by dan-lior** [reply]
I just naively added the line 

bdry_info.parallel_sync_side_ids();

right after the loop above and it seems to have solved the problem. 

Thanks John! 

PS Not to seem ungrateful, but I searched for just such a synchronization function in the online documentation for the BoundaryInfo class and didn't find it. Perhaps it comes from a parent class? 

dan

**Reply by jwpeterson** [reply]
https://github.com/libMesh/libmesh/blob/0f2c5cfcad0b8d8984dd5a3ddc678eb352152995/include/mesh/boundary_info.h#L722-L731

Most likely the issue is that our Doxygen documentation is not up-to-date for this class, so if you searched for it there, you would not have found it. I would recommend using the Doxygen docs maintained by the MOOSE project, [BounadryInfo](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1BoundaryInfo.html), as they are automatically updated on a more regular basis.
